src/ui/Base/Initialization.ts

```diff
@@ -242,41 +242,66 @@ export class Initialization {
     const alreadyInitialized = Component.get(element, QueryController, true);
     if (alreadyInitialized) {
       this.logger.error('This DOM element has already been initialized as a search interface, skipping initialization', element);
-      return new Promise((resolve, reject) => {
-        resolve({ elem: element });
-      });
+      return {
+        elem: element
+      };
     }
 
     options = Initialization.resolveDefaultOptions(element, options);
 
+    const waitForAllPromisesToFinish = async (eventType: string, promises: Promise<any>[]) => {
+      const promisesWithErrorsHandledIndividually = promises.map(p => {
+        return p.catch(error =>
+          this.logger.warn(
+            `An error occurred when trying to defer the \"${eventType}\" event. The defer will be ignored.`,
+            `Error: ${error}`
+          )
+        );
+      });
+      return await Promise.all(promisesWithErrorsHandledIndividually).catch(error =>
+        this.logger.error(
+          `An unexpected error occurred when trying to defer the \"${event}\" event. All defers will be ignored.`,
+          `Error: ${error}`
+        )
+      );
+    };
+
+    const triggerInitializationEventWithArguments = async (eventType: string) => {
+      const initializationEventArgs: IInitializationEventArgs = {
+        defer: []
+      };
+      $$(element).trigger(eventType, initializationEventArgs);
+      await waitForAllPromisesToFinish(eventType, initializationEventArgs.defer);
```

Comment:
We should be checking that the property `initializationEventArgs.defer` contains something, the reasoning being that with your pull requests, this breaks a lot of backward compatibility since it means all the initialization events now have a different signature (they all return a promise, instead of void when we do non-lazy init).

There should not be any need to modify any of the existing unit tests that we have currently, since it's an optional parameter that users can now decide to use.

If they do decide to use it, then it's fair to assume that they need to adapt their code (because it is optional).